RFC-0009-jdbc-join-push-down.md

```diff
@@ -0,0 +1,798 @@
+                buildFunctionTranslator(ImmutableSet.of(JoinOperatorTranslators.class)),
+                identifierQuote);
+```
+We have also added visitPlan() method and enhanced the visitFilter() method.
+
+```
+@Override
+        public PlanNode visitPlan(PlanNode node, Void context)
+        {
+            ImmutableList.Builder<PlanNode> children = ImmutableList.builder();
+            boolean changed = false;
+            for (PlanNode child : node.getSources()) {
+                PlanNode newChild = child.accept(this, null);
+                if (newChild != child) {
+                    changed = true;
+                }
+                children.add(newChild);
+            }
+
+            if (!changed) {
+                return node;
+            }
+            return node.replaceChildren(children.build());
+        }
+```
+
+Enhancements done to visitFilter are from this PR : https://github.com/prestodb/presto/pull/16412/
+Before, if all the filter could not be pushed down, nothing would be pushed down.
+The changes help in finding what all can be pushed down and able to be translated. The filters that can't be pushed down are kept in a new FilterNode.
+
+## Self Join
+Self join is a scenario where a table is joined with itself. This works natively in Presto. When you do a self join in presto, that is sent as select statements to the datasource first. On top of that presto does all the joining and filtering etc. 
+We want to push down this self join to database level. This is similar to pushing down an inner join. But, we faced some difficulties in achieving this.
+
+#### Difficulty faced :
+If we do a join query which joins the same table with itself, We get 2 TableScans in the plan.
+There isn’t a way to differentiate between the columns of these TableScans.
+
+Example : Assume there is a column 'col_1' in a table 'table_1'.  We join table_1 with itself ->
+```
+select * from table_1 a join table_1 b on a.col_1 = b.col_1;
+```
+The column 'col_1' will be referred as col_1 (in outputVariables) inside TableScan1 and will point to 'col_1' JdbcColumnHandle (in assignments)
+It will be referred as col_1_0 (in outputVariables) inside TableScan2 and will point to 'col_1' JdbcColumnHandle (in assignments)
+When our Optimiser 'JdbcJoinRenderByConnector' is run, both these TableScans get converted to a Single TableScan, with assignments, outputVariables, etc. combined.
+(Note : assignments now has two references to the same column 'col_1'. One from 'col_1' and the other from 'col_1_0')
+This creates a problem in pushPredicateIntoTableScan method in PickTableLayout class, which assumes that all the values in assignments are unique.
+This assumption by presto is done on the basis that till now all queries done by presto to any database is as a select from a single table.
+
+#### Proposed Solution :
+We can add a field 'table_alias' (table alias) to JdbcTableHandle. This is not the table alias that the user passes in his query.
```

Comment:
Please get details from latest implementation and add here.

Reply:
added